dipy/reconst/weights_method.py

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/python
+""" Functions for defining weights for iterative fitting methods."""
+
+import numpy as np
+
+MIN_POSITIVE_SIGNAL = 0.0001
+
+
+def simple_cutoff(residuals, log_residuals, pred_sig,
+                  design_matrix, leverages, C, cutoff):
+    """ Define outliers based on signal.
+    """
+    leverages[np.isclose(leverages, 1.0)] = 0.99  # avoids rare issues
+    HAT_factor = np.sqrt(1 - leverages)
+    cond = (np.abs(residuals) > + cutoff * C * HAT_factor)
+    return cond
+
+
+def two_eyes_cutoff(residuals, log_residuals, pred_sig,
+                    design_matrix, leverages, C, cutoff):
+    """ Two-eyes approach to define outliers, see [1]_.
+
+    References
+    ----------
+    .. [1] Collier, Q., Veraart, J., Jeurissen, B., den Dekker, A., Sijbers,
+       J., 2015. Iterative reweighted linear least squares for accurate, fast,
+       and robust estimation of diffusion magnetic resonance parameters: IRLLS
+       for Estimation of Diffusion MR Parameters. Magnetic Resonance in
+       Medicine 73, 2174-2184.
+    """
+    leverages[np.isclose(leverages, 1.0)] = 0.99  # avoids rare issues
+    HAT_factor = np.sqrt(1 - leverages)
+    cond = (residuals > + cutoff * C * HAT_factor) |\
+           (log_residuals < - cutoff * C * HAT_factor / pred_sig)
+    return cond
+
+
+def weights_method_wls_m_est(data, pred_sig, design_matrix, leverages,
+                             idx, total_idx, last_robust,
+                             m_est="gm", cutoff=3,
+                             outlier_condition_func=simple_cutoff):
+    """ Calculate M-estimator weights for WLS model.
+
+    Parameters
+    ----------
+    data : array
+        The measured signal.
+    pred_sig : array
+        The predicted signal, given a previous fit.
+        Has the same shape as data.
+    design_matrix : array (g, ...)
+        Design matrix holding the covariants used to solve for the
+        regression coefficients.
+    leverages : array
+        The leverages (diagonal of the 'hat matrix') of the fit.
+    idx : int
+        The current iteration number.
+    total_idx : int
+        The total number of iterations.
+    last_robust : array (int),
+        Array of 1s (inlier) and 0s (outlier).
+        Has the same shape as data.
+    m_est : which M-estimator weighting scheme to use. Currently,
+            'gm' (Geman-McClure) and 'cauchy' are provided.
+    cut_off : float
+        Outliers are defined by residuals > |cut_off * standard_deviation|
+
```

Comment:
missing one arguments

Reply:
Check how this renders:
- Every time Sphinx finds an asterisk it thinks that an emphasized word starts (cannot say out the top of my head if a white space before the following word avoids this)
- May yield an `Undefined substitution referenced: "cut_off * standard_deviation".`: https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#substitution-references

Reply:
replaced "*" with "x" for multiplication, looks better anyway, thanks for the spot